### ISSUE TITLE
Rule Change: Support nested object paths in no-restricted-properties

### ISSUE THREAD

**Issue by sethamus** [eslint/eslint#19809]
### What rule do you want to change?

no-restricted-properties

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
/* eslint no-restricted-properties: [2, {
    "object": "config.apiKey",
    "allowProperties": ["settings"]
}] */

config.apiKey.settings = "12345";
config.apiKey.other = "value";
```

### What does the rule currently do for this code?

The rule currently doesn't report any errors.

### What will the rule do after it's changed?

```js
config.apiKey.settings = "12345"; // No error
config.apiKey.other = "value"; // Error: 'config.apiKey.other' is restricted from being used
```

### Participation

- [x] I am willing to submit a pull request to implement this change.

**Comment by nzakas** [comment]
Thanks for the suggestion. I'm :-1: as I feel like this rule already does everything that it should. Starting to go down nested object paths seems like added complexity for very little gain and ever-increasingly an edge case.

Let's see what the rest of @eslint/eslint-team thinks.

**Comment by nzakas** [comment]
@eslint/eslint-team still looking for feedback here.

**Comment by snitin315** [comment]
I agree with @nzakas on this

**Comment by Tanujkanti4441** [comment]
I also agree with the team.

```js
/* eslint no-restricted-properties: [2, {
    "object": "config.apiKey",
    "allowProperties": ["settings"]
}] */

config.apiKey.settings = "12345";
config.apiKey.other = "value";
```

Since now we have `allowProperties` option we can achieve the expected behavior by updating the code as:

```js
/* eslint no-restricted-properties: [2, {
    "object": "config",
    "allowProperties": ["apiKey", "settings"]
}] */

const config = config.apiKey;

config.settings = "12345"; // No error
config.other = "value"; // error
```

Should we add such example in docs to show this can be done for nested properties?

**Comment by mdjermanovic** [comment]
> 
> /* eslint no-restricted-properties: [2, {
>     "object": "config.apiKey",
>     "allowProperties": ["settings"]
> }] */
> 
> config.apiKey.settings = "12345";
> config.apiKey.other = "value";
> Since now we have `allowProperties` option we can achieve the expected behavior by updating the code as:
> 
> /* eslint no-restricted-properties: [2, {
>     "object": "config",
>     "allowProperties": ["apiKey", "settings"]
> }] */
> 
> const config = config.apiKey;
> 
> config.settings = "12345"; // No error
> config.other = "value"; // error
> Should we add such example in docs to show this can be done for nested properties?

I'm not sure if an example of how the code can be rewritten in a particular way so that the rule can catch the problem would be generally useful. In this example, I think it's much more likely that the variable name would be `apiKey` (`const apiKey = config.apiKey;`) than `config`.

**Comment by mdjermanovic** [comment]
I agree with @nzakas that this seems out of scope for this rule. This rule is already based on the assumption that certain variable names will always be used for certain objects, but restricting nested paths would be much less reliable, as the rule would be able to disallow only direct access, like `config.apiKey.other = "value";`, but not, for example, `const apiKey = config.apiKey; apiKey.other = "value";`.

As it looks like the team is not in favor of this proposal, I'm closing it.
